core/src/main/scala/kafka/network/SocketServer.scala

```diff
@@ -974,8 +973,39 @@ private[kafka] class Processor(val id: Int,
                 val context = new RequestContext(header, connectionId, channel.socketAddress,
                   channel.principal, listenerName, securityProtocol,
                   channel.channelMetadataRegistry.clientInformation, isPrivilegedListener)
-                val req = new RequestChannel.Request(processor = id, context = context,
-                  startTimeNanos = nowNanos, memoryPool, receive.payload, requestChannel.metrics)
+
+                val principalSerde = Option(channel.principalSerde.orElse(null))
+                val req =
+                if (header.apiKey == ApiKeys.ENVELOPE) {
```

Comment:
nit: this is misaligned. It might be better to pull the body here into a separate method (e.g. `parseEnvelopeRequest`)